### DISCUSSION TITLE
V5 & Monorepo: Error: Trans component was rendered without I18nProvider

### DISCUSSION THREAD

**Discussion by francois-pasquier** [lingui/js-lingui#2091] [Q&A]
Hey, I will try to provide a monorepo but for the moment, I'll just hope someone has an idea:

I started from the lingui `next` branch with the `nextjs-swc` example.

In the context of a monorepo,  have a nextjs project in `apps/web`. Inside of it, I have this as a dependency: `"@lingui/react": "5.0.0-next.4",`

I also have a shared codebase in `packages/app`, inside of it I have basically all of my code, nextjs only loads the screens from `packages/app`.
In order for my `<Trans />` calls to work, I had to declare `"@lingui/react": "5.0.0-next.4",` as a dependency as well.

The issue is, when you declare twice `"@lingui/react": "5.0.0-next.4",` in a monorepo you end up with this error:

```
Error: Trans component was rendered without I18nProvider.
Attempted to render message: is this going to be translated? id: 3FOdsM. Make sure this component is rendered inside a I18nProvider.
```

It honestly took me hours to troubleshoot this one.

For the moment I have the `@lingui/react` package declared in my root package.json. It's not a proper fix though.

**Comment by timofei-iatsenko** [comment]
The doing dependency right in the monorepo is a complicated topic. It's not related to lingui, but to the monorepos itself. The main rule is, if you import something in the package, you should have this stated in the dependency. If all of your code lies in the `packages/app` and have imports from `@lingui/react` you should state it in this package. If your `apps/web` also has any imports from `@lingui/react` it should also declare it in the direct dependencies, even if it's already declared in one of the subdependencies. 
You can use [import/no-extraneous-dependencies](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-extraneous-dependencies.md) eslint rule to quickly spot this issues. 

Regarding your issue, first you need to check that you don't have few versions of `@lingui/react` package installed, the `yarn why` or `pnpm why` command will help you with that. 

Next, because I don't know how your monorepo works, it might be the case that you bundle your `packages/app` with `@lingui/react` inside, so you again end up with 2 copies of the package. 

The same could be with a React package, make sure it with a `yarn why` command that you don't have two copies of package installed. Again, you have to have React be stated in the direct dependencies of the both `packages/app`  and `apps/web` because they both uses react, and they should have exactly the same version. 

Also pay attention to the `peerDependencies` issues when installing the packages, it usually shows when the peer dependencies are incorrectly meet which could potentially lead to this kind of issues.

**Comment by devpla** [comment]
I ran into this exact issue in a pnpm + monorepo setup with the following structure:

```
apps/my-project      ← next.js frontend
packages/components  ← shared UI components (includes calls)
```

Even though I had all the right Lingui setup (one single `LinguiClientProvider`, proper `peerDependencies`, `transpilePackages`, etc.), I was still getting:

> Error: Trans component was rendered without I18nProvider

After hours of debugging, I finally traced it down to this:  
 **`@lingui/react` was being installed twice with different peer resolution hashes**, even though the version was identical.

You can confirm it via:

```js
require.resolve('@lingui/react'))
```

In my case, I saw two different .pnpm/ hash paths due to different typescript versions between apps/my-project and packages/components. 

<img width="1784" height="148" alt="image" src="https://github.com/user-attachments/assets/0de9d53f-4e91-4ac5-b953-c7ccfb245bf8" />

Once I aligned the TypeScript versions across both packages, the issue disappeared completely.

**Ensure typescript, react, and `@lingui/*` peer dependencies are perfectly aligned across all packages.**
Additionally, adding the following overrides in the root package.json may help prevent further divergence:
```
  "pnpm": {
    "overrides": {
      "@lingui/react": "5.4.1",
      "@lingui/core": "5.4.1",
      "react": "19.0.0",
      "typescript": "5.7.3"
    }
  }
```

Hope this saves someone a few hours of headbanging.  🙏

**Reply by dustin-ruetz** [reply]
Thanks for this explanation, @devpla! It helped me resolve a similar error in my Lingui monorepo 🙏

Just wanted to add that pnpm's [catalogs](https://pnpm.io/catalogs) feature for workspaces is nicely-suited for solving this problem in monorepos as well, since it allows you to define exact package versions in the `pnpm-workspace.yaml` file at the root of the repo and then reference them via the `catalog:` protocol in all of monorepo's subpackages. Abridged code excerpts below:

```yaml
# pnpm-workspace.yaml
packages:
  - apps/*
  - packages/*

catalog:
  '@lingui/core': 5.6.1
  '@lingui/react': 5.6.1
  '@types/react': 19.2.7
  '@types/react-dom': 19.2.3
  react: 19.2.1
  react-dom: 19.2.1
  typescript: 5.9.3
```

```json
// packages/ui/package.json
{
  "name": "@packages/ui",
  "devDependencies": {
    "@lingui/core": "catalog:",
    "@lingui/react": "catalog:",
    "@types/react": "catalog:",
    "react": "catalog:",
    "typescript": "catalog:"
  },
  "peerDependencies": {
    "@lingui/core": "catalog:",
    "@lingui/react": "catalog:",
    "@types/react": "catalog:",
    "react": "catalog:",
    "typescript": "catalog:"
  }
}
```

```json
// apps/app-1/package.json
{
  "name": "@apps/app-1",
  "dependencies": {
    "@lingui/core": "catalog:",
    "@lingui/react": "catalog:",
    "@packages/ui": "workspace:*",
    "react": "catalog:",
    "react-dom": "catalog:",
  },
  "devDependencies": {
    "@types/react": "catalog:",
    "@types/react-dom": "catalog:",
    "typescript": "catalog:"
  }
}
```
